### DISCUSSION TITLE
Help

### DISCUSSION THREAD

**Discussion by Sinfolke** [ASDAlexander77/TypeScriptCompiler#43] [General]
Could i help you with the project? I don't know c++ but i know ts. For example i can try to create a configuration that translates ts methods to c++ methods since i noticed it does not translate methods.

**Comment by ASDAlexander77** [comment]
Thank you for your interest in the project.  Not sure yet which help I would need in that project.  But I am thinking that next would be good to have "WASM" tests. so you can try to use it as WASM compiler and see what is missing from point of usability.

**Reply by Sinfolke** [reply]
Did you want if i would test the compiler in goldbold

**Reply by ASDAlexander77** [reply]
yes, it is fine, but it doesn't have the latest version so you may see the issue which already fixed.

**Reply by Sinfolke** [reply]
I tested and it commonly works well. However 
It does not compile code with methods correctly. Some methods (for example lenght in string) works because has the same name and implementation in c++. I have already created an example of tool by which we can implement configuration to tranform ts methods and properties to c++. I don't know whether you have already implemented it, but i will attack it here when will be able

**Comment by ASDAlexander77** [comment]
your repo is not publicly accessible. can't open it

**Reply by Sinfolke** [reply]
It is public now

The builden configuration:

```
{
    "String":{
        "at":{
            "type":"partialTransformMethod",
            "code":"char __String__at(std::string current, int index) {\r\n    if (index < 0) {\r\n        std::reverse(current.begin(), current.end());\r\n        index = index * -1 - 1;\r\n    }\r\n    if (index > current.length() - 1) {\r\n        throw \"Index is out of bound\";\r\n    }\r\n    return current.at(index);\r\n}\n",
            "include":["string","algorithm"]
        },
        "charAt":{
            "type":"partialTransformMethod",
            "code":"char __String__charAt(std::string current, std::optional<int> index) {\r\n    int id = index.value();\r\n    if (!index.has_value()) {\r\n        id = 0;\r\n    } else if (id < 0 || id > current.length() - 1) {\r\n        return '\\0';\r\n    }\r\n    return current.at(id);\r\n}\n",
            "include":["string","optional"]
        },
        "charCodeAt":{
            "type":"partialTransformMethod",
            "code":"int __String__charCodeAt(const std::string current, std::optional<int> index) {\r\n    const int id = index.has_value() ? index.value() : 0;\r\n    if (id < 0 || id >= current.length()) {\r\n        return '\\0';\r\n    }\r\n    unsigned char charValue = static_cast<unsigned char>(current[id]);\r\n    return static_cast<int>(charValue);\r\n}\n",
            "include":["iostream","string","optional"]
        },
        "codePointAt":{
            "type":"partialTransformMethod",
            "lib":{
                "https://github.com/nemtrif/utfcpp.git": {
                    "main":"source/utf8.h",
                    "include":[
                        "source/utf8.h","source/utf8"
                    ]
                }
            },
            "code":"// also required library utf-8\r\nint __String__codePointAt(const std::string& str, std::optional<size_t> index) {\r\n    // possibly undefined (no value) so replaced to 0 according\r\n    // to documentation\r\n    const int value = index.has_value() ? index.value() : 0; \r\n    if (value >= str.length()) {\r\n        // invalid index\r\n        return '\\0';\r\n    }\r\n\r\n    std::string::const_iterator it = str.begin();\r\n    std::string::const_iterator end = str.end();\r\n    size_t currentIndex = 0;\r\n\r\n    while (it != end && currentIndex < value) {\r\n        utf8::next(it, end);\r\n        ++currentIndex;\r\n    }\r\n\r\n    if (it != end) {\r\n        utf8::uint32_t codePoint = utf8::next(it, end);\r\n        return static_cast<int>(codePoint);\r\n    }\r\n    // Invalid index or invalid UTF-8 sequence\r\n    return '\\0';\r\n}\n",
            "include":[
                "iostream",
                "string",
                "optional"
            ]
        }
    }
}
```

**Reply by ASDAlexander77** [reply]
this would be useful for TypeScript2C++ but I am working on TypeScriptCompiler now

**Reply by Sinfolke** [reply]
Ok, you might can somehow implement the c++ code at src/String/cpp folder since your code is written at c++. But if not possible then im sorry 😅

**Reply by ASDAlexander77** [reply]
cool thing about TypeScript Native compiler that I do not need to use C++ anymore :) .. the only thing is just SDKs which are written in C++ such as Vulkan etc

**Reply by Sinfolke** [reply]
I honestly know c++ only a little bit and i know only the main basic of the language. If i can somehow help you with project (like a tried to write js methods on c++ saving same behaviour) then please, tell. 

I tested compiler on web and found some issues
1. the constructor is considered method of class. And so it can be called:
```
const instance = new myClass();
// error in js but valid using your compiler
instance.constructor()
```

2. The compiler is not able to access a key of object by variable:
```
const obj = {
key1: "value1",
key2: "value2",
}
const accessKey = "key1";
function main() {
     // compiler error
     print(obj[accessKey]);
}
```
3. Has some issues with outputting array or object
```
    const dynamicArr = [
        1,
        "first",
        [
            "second",
            [
                'third',
            ]
        ]
    ];
    
    function main() {
        /*  
            prints 
    
    
            print([1, 2, 3]);
    
            throws error
            print(dynamicArr)
        */
}
```
4. Has issues with null (program crashes)
5. In a string escaptions (like \n, \t...) Do not work
Everything else is seems good

**Reply by ASDAlexander77** [reply]
print can print only simple data, for printing array u need to write the code for it

**Reply by ASDAlexander77** [reply]
regarding 2.  it is working like C++, u can access field but "variable" only static access via field name
